core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala

```diff
@@ -0,0 +1,175 @@
+
+  addLogAppender()
+
+  private def addLogAppender(): Unit = {
+    val appenders = LogManager.getRootLogger().getAllAppenders()
+    val layout = if (appenders.hasMoreElements()) {
+      appenders.nextElement().asInstanceOf[Appender].getLayout()
+    } else {
+      new PatternLayout(conf.get(DRIVER_LOG_LAYOUT))
+    }
+    val fa = new Log4jFileAppender(layout, localLogFile)
+    fa.setName(DriverLogger.APPENDER_NAME)
+    LogManager.getRootLogger().addAppender(fa)
+    logInfo(s"Added a local log appender at: ${localLogFile}")
+  }
+
+  def startSync(hadoopConf: Configuration): Unit = {
+    try {
+      // Setup a writer which moves the local file to hdfs continuously
+      if (syncToDfs) {
+        val appId = Utils.sanitizeDirName(conf.getAppId)
+        writer = Some(new HdfsAsyncWriter(appId, hadoopConf))
+      }
+    } catch {
+      case e: Exception =>
+        logError(s"Could not sync driver logs to spark dfs", e)
+    }
+  }
+
+  def stop(): Unit = {
+    try {
+      writer.map(_.closeWriter())
+    } catch {
+      case e: Exception =>
+        logError(s"Error in persisting driver logs", e)
+    } finally {
+      try {
+        JavaUtils.deleteRecursively(FileUtils.getFile(localLogFile).getParentFile())
+      } catch {
+        case e: Exception =>
+          logError(s"Error in deleting local driver log dir", e)
+      }
+    }
+  }
+
+  private class HdfsAsyncWriter(appId: String, hadoopConf: Configuration) extends Runnable
+    with Logging {
+
+    private var streamClosed = false
+    private val inStream = new BufferedInputStream(new FileInputStream(localLogFile))
+    private var hdfsLogFile: String = {
+      val rootDir = conf.get(DRIVER_LOG_DFS_DIR.key, "/tmp/driver_logs").split(",").head
+      FileUtils.getFile(rootDir, appId, DRIVER_LOG_FILE).getAbsolutePath()
+    }
+    private var fileSystem: FileSystem = FileSystem.get(hadoopConf)
+    private val outputStream: FSDataOutputStream = fileSystem.create(new Path(hdfsLogFile), true)
+    fileSystem.setPermission(new Path(hdfsLogFile), LOG_FILE_PERMISSIONS)
+    private val tmpBuffer = new Array[Byte](UPLOAD_CHUNK_SIZE)
+    private val threadpool = ThreadUtils.newDaemonSingleThreadScheduledExecutor("dfsSyncThread")
+    threadpool.scheduleWithFixedDelay(this, UPLOAD_INTERVAL_IN_SECS, UPLOAD_INTERVAL_IN_SECS,
+      TimeUnit.SECONDS)
+    logInfo(s"The local driver log file is being synced to spark dfs at: ${hdfsLogFile}")
+
+    def run(): Unit = {
+      if (streamClosed) {
+        return
+      }
+      try {
+        var remaining = inStream.available()
+        while (remaining > 0) {
+          val read = inStream.read(tmpBuffer, 0, math.min(remaining, UPLOAD_CHUNK_SIZE))
+          outputStream.write(tmpBuffer, 0, read)
+          remaining -= read
+        }
+        outputStream.hflush()
+      } catch {
+        case e: Exception => logError("Failed to write to spark dfs", e)
+      }
+    }
+
+    def close(): Unit = {
+      try {
```

Comment:
Return early if `streamClosed`?

Reply:
This should not be called twice, so it is not needed.

Reply:
It's good practice to write idempotent close methods. "Should not be called twice" is part of the famous last words list.